Effect enable switches changes

Be-ing opened the pull request:
This PR disables the per-effect switches when Mixxx starts and makes the switches to enable EffectUnits on decks (and other sources) persistent.

The motivation for this came from my experience setting up for my set last night using the persistent effects in PR #1092. I was worried because the level meters on my sound card showed no signal was coming out. I had to ask the previous DJ to play another track while I figured out what was going wrong. I thought something was wrong with my sound card. I tried restarting Mixxx and replugging my sound card several times and restarted my computer. Finally, I noticed that the level meters on Mixxx were not showing any signal coming through when the deck was playing. I realized that the dry/wet knob for the EffectUnit enabled on that deck was all the way wet and the filter effect's metaknob was all the way right, completely silencing the deck. I have my mapping set up to set the initial position of the software knobs to the position of the hardware knobs on startup, so the knobs may have accidentally been turned that way during transit to my gig. With this PR, that situation would not happen because each effect would have to be explicitly enabled by the user when they want to use it.

Making the deck enable switches persistent allows controllers lacking buttons to control which effect units are active on which decks to use effects conveniently. This approach is better than requiring the controller's mapping to set the enable switches in its init() function because the user may want to change their defaults and change the switches throughout a set. The user would have to enable the effect units on their preferred decks once and could leave them that way indefinitely if they choose, but users who want otherwise won't have their preferences overriden by the controller mapping's init() function.

Be-ing committed:
turn off effects by default

Be-ing committed:
make EffectUnit deck assignment switches persistent

daschuer commented:
I like the idea. Unfortunately we cannot merge it without changing the way our default LateNight skin works.

Be-ing committed:
LateNight: quick hack to not break effects

Be-ing commented:
The last commit should hold over LateNight until someone takes up the work of redesign its effects UI.

Be-ing committed:
fix effect enable buttons on startup

Be-ing committed:
fix EffectUnit channel asignment buttons on startup

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -147,10 +147,6 @@ void EffectSlot::loadEffect(EffectPointer pEffect) {
-        // Enabled is a persistent property of the effect slot, not of the
-        // effect. Propagate the current setting to the effect.
-        pEffect->setEnabled(m_pControlEnabled->get() > 0.0);
-
```
This is a no-op with the initial value of `m_pControlEnabled` equal to the initial value of Effect's `m_bEnabled`.

Be-ing committed:
make effects loaded after initialization load with the correct enabled/disabled state

Be-ing commented:
@timrae : Could you test this? I have it merged into my personal branch. I think it's ready for merge.

timrae commented:
Sure I'm already back to work but I'll try test it over the next few days.

timrae commented:
Did you figure out the steps to reproduce? I'm not exactly sure how to test it

timrae commented:
Thanks, I can confirm that yes that was the bug that I reported, and yes it is fixed here. I'm also 100% in support of disabling the individual FX enable switches and making the assignments persistent. It looks good for the merge to me.

rryan commented on src/effects/effectchainslot.cpp:
```diff
@@ -269,14 +269,16 @@ void EffectChainSlot::registerChannel(const ChannelHandleAndGroup& handle_group)
     m_channelStatusMapper.setMapping(pEnableControl, handle_group.name());
     connect(pEnableControl, SIGNAL(valueChanged(double)),
             &m_channelStatusMapper, SLOT(map()));
+    slotChannelStatusChanged(handle_group.name());
```
We already have pInfo here so you could avoid the QMap lookup by doing a 

```
if (pEnableControl->toBool()) {
  m_pEffectChain->enableForChannel(pInfo->handle_group);
} else {
  m_pEffectChain->disableForChannel(pInfo->handle_group);
}
```

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -146,9 +147,7 @@ void EffectSlot::loadEffect(EffectPointer pEffect) {
-        // Enabled is a persistent property of the effect slot, not of the
-        // effect. Propagate the current setting to the effect.
-        pEffect->setEnabled(m_pControlEnabled->get() > 0.0);
+        pEffect->updateEngineState();
```
Couldn't the effect's enabled state be out of sync with the EffectSlot's enabled status? That's what the original line was doing here.

Be-ing replied:
No, the original line was a no-op because setEnabled() checks if the parameter passed to it is equal to its private enabled boolean. pEffect->updateEngineState() is necessary to sync the state of the EffectSlot and the Effect when a new effect is loaded to an enabled effect slot now that Effects are disabled by default.

rryan replied:
If the effect slot is enabled then the effect will be disabled but the effect slot will be enabled.

It should be easy to see -- enable an effect, then change the effect to something else. The enabled button will be lit but the effect will be disabled.

Be-ing replied:
The problem I was fixing was that when the effect slot was disabled, loading a new effect would load the EngineEffect enabled but the slot would still be disabled.

rryan replied:
Ah, looks like you also need to make EngineEffect default to disabled so Effect/EngineEffect don't start life out of sync. I think this line should stay as it was.

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -31,9 +31,10 @@ EffectSlot::EffectSlot(const QString& group,
-    m_pControlEnabled->set(true);
+    // Default to disabled to prevent accidental activation of effects
+    // at the beginning of a set.
+    m_pControlEnabled->setDefaultValue(false);
```
These are both no-ops since they're the default. Maybe drop them and move the comment above the "new ControlPushButton".

rryan commented:
Failing test:

```
[ RUN      ] EffectSlotTest.ControlsReflectSlotState

Loading resources from  "/Users/travis/build/mixxxdj/mixxx/res/"

src/test/effectslottest.cpp:62: Failure

Value of: ControlObject::get(ConfigKey(group, "enabled"))

  Actual: 0

Expected: 1.0

Which is: 1

src/test/effectslottest.cpp:68: Failure

Value of: pEffect->enabled()

  Actual: false

Expected: true

"[EffectRack1_EffectUnit1_Effect1]" , "loaded" is read-only. Ignoring set of value: 0

"[EffectRack1_EffectUnit1_Effect1]" , "num_parameters" is read-only. Ignoring set of value: 2

[  FAILED  ] EffectSlotTest.ControlsReflectSlotState (72 ms)
```

rryan commented:
I think it's inconsistent to default the EffectChainSlot to enabled but the EffectSlot to disabled or vice versa. They were both enabled by default to support skins that don't show either one of them with little hassle. Let's disable them both by default and then put the work onto the skins to set things up according to the paradigm the skin offers.

rryan commented on res/skins/LateNight/effect_unit.xml:
```diff
@@ -93,11 +93,12 @@
                   <Text>ON</Text>
                 </State>
                 <Connection>
-                  <ConfigKey persist="true"><Variable name="group"/>,enabled</ConfigKey>
```
I'm boggling a bit as to why there were two "enabled" connections here. @ywwg do you remember?

Be-ing replied:
O.o I don't understand either. This whole `<Connection>` element should probably be removed.

rryan replied:
Yea, I think it probably can be.

rryan replied:
Since this is empty now could you remove it?

Be-ing commented:
Let's stop providing an inconsistent interface in different skins. Skins should conform to a standard interface or be expected to not work properly. I'm leaning towards removing the chain enable switches from skins. It's one more layer of complexity to get confused about while mixing. Effect enable and deck assignment switches is already pretty complicated.

Be-ing commented:
What is the use case for chain enable switches separate from deck assignment switches?

rryan commented:
> What is the use case for chain enable switches separate from deck assignment switches?

They let you drop an effect unit in and out of the mix with one click -- as LateNight does.

Be-ing commented:
> They let you drop an effect unit in and out of the mix with one click -- as LateNight does.

So do effect enable switches and deck assignment switches. But, confusingly, those won't work if the whole chain is disabled.

timrae commented:
> I think it's inconsistent to default the EffectChainSlot to enabled but the EffectSlot to disabled or vice versa.

Having spent time mixing with the new branch, I strongly disagree with this. Having the states of the individual enabled switches to "off" makes sure that the unit doesn't affect the sound by default. Going further to disable the unit itself would make it such that users who are experimenting with effects for the first time have to press THREE buttons before they can use the FX:

channel assignment, effect enable, effect unit enable.

I would go further to say that by default only ONE button press should be required by default: the effect enable. I.e. channel assignment should default to 1 and 2 for the two units respectively

rryan commented:
> They let you drop an effect unit in and out of the mix with one click -- as LateNight does.
> So do effect enable switches and deck assignment switches. But, confusingly, those won't work if the whole chain is disabled.

Right -- in practice the skin supports either enabling at the chain level or the effect level.

rryan commented:
> Going further to disable the unit itself would make it such that users who are experimenting with effects for the first time have to press THREE buttons before they can use the FX:

I think you misunderstood the proposal -- we would default all of them to off in C++ and then put it on the skin to set one of the two to on via an attributes section.

Be-ing commented:
Skins shouldn't have to worry about a ControlObject that they do not show.

rryan commented:
> Skins shouldn't have to worry about a ControlObject that they do not show.

Then let's keep the status quo (all enabled by default) and have the skin disable the units they show by default?

Be-ing commented:
> Then let's keep the status quo (all enabled by default) and have the skin disable the units they show by default?

No, then controller users have to click a button on screen to start using effects. AFAIK no controllers are designed for chain enable switches separate from deck assignment and effect enable switches.

timrae commented:
> I think you misunderstood the proposal -- we would default all of them to off in C++ and then put it on the skin to set one of the two to on via an attributes section.

Ah yes I did, thanks for the clarification. In that case I don't have a strong opinion, though I feel it should be a requirement for all skins to have the behavior I described, in which case why not enforce it at the C++ layer instead of requiring skin creators to implement this.

timrae commented:
How about syncing the state of the chain with the individual units? Switching the unit off switches all fx off

timrae commented:
And whenever one or more effects are on, the chain is on

Be-ing commented:
> How about syncing the state of the chain with the individual units? Switching the unit off switches all fx off

IMO that would be overcomplicated and unnecessary. The same can be accomplished by unassigning a chain from a deck.

timrae commented:
True. In that case just remove it.

Be-ing commented:
> channel assignment should default to 1 and 2 for the two units respectively

Hmmm, how does this fit with making the deck assignments persistent? Should deck assignments be persistent, or should there always be a 1:1 setup between effect units and decks on startup? Maybe assign effect units to the corresponding deck if all the deck assignments are disabled on startup?

timrae commented:
I'd go with the simplest option: make it persistent, but give default values of channel 1 and channel 2 in the case where the user has not selected anything. I don't think there's a need to get too "clever" and reassign them if the user disabled them previously.

Be-ing commented:
The trouble is distinguishing the cases when a user is starting Mixxx for the first time or upgrading from an old version versus the user has intentionally unassigned an EffectUnit from all decks.

Be-ing commented:
Maybe that could be handled in src/preferences/upgrade.cpp? That seems like an ugly solution though. @rryan , thoughts?

rryan commented:
>> Then let's keep the status quo (all enabled by default) and have the skin disable the units they show by default?
> No, then controller users have to click a button on screen to start using effects. 

Sorry, I used the wrong term -- I didn't mean "units" I meant "effects". What I meant was the unit would be enabled by default (as it is now) and the skin would disable the effect slots at start. In this case, the controller wouldn't have to touch the unit's enabled control.

However...

> AFAIK no controllers are designed for chain enable switches separate from deck assignment and effect enable switches.

Hm good point -- nevermind then. The inconsistency of enabling the effect chain slot by default and disabling the effect slot by default is fine w/ me.

> How about syncing the state of the chain with the individual units? Switching the unit off switches all fx off. And whenever one or more effects are on, the chain is on

This would be nice but it's surprisingly complicated to keep multiple controls in sync code-wise and it's easy to create infinite loops. Since the control system isn't transactional we can't update multiple controls atomically, and when you change one, it can run change handlers that go and change the state of the world on you before you update the other one to be in sync.

rryan commented:
> The trouble is distinguishing the cases when a user is starting Mixxx for the first time or upgrading from an old version versus the user has intentionally unassigned an EffectUnit from all decks.

I think a good fix would be to support providing the control's default value in the ControlObject/ControlDoublePrivate constructor. Right now we default to zero if persist is true and the value is not present or malformatted in the user's config:
https://github.com/mixxxdj/mixxx/blob/master/src/control/control.cpp#L53

Be-ing commented:
SGTM. Should that be in a separate PR or together with this?

timrae commented:
> The trouble is distinguishing the cases when a user is starting Mixxx for
the first time or upgrading from an old version versus the user has
intentionally unassigned an EffectUnit from all decks.

> Maybe that could be handled in src/preferences/upgrade.cpp?

That's exactly how we handle such issues in other projects that I've worked
on. You give a default value when you pull it out of the QSettings object
to handle the case where users are setting it for the first time, and you
force the value in the upgrade code whenever the value was set in a
previous version and needs to be changed.

On 5 Jan. 2017 07:07, "RJ Skerry-Ryan" <notifications@github.com> wrote:

> Then let's keep the status quo (all enabled by default) and have the skin
> disable the units they show by default?
> No, then controller users have to click a button on screen to start using
> effects.
>
> Sorry, I used the wrong term -- I didn't mean "units" I meant "effects".
> What I meant was the unit would be enabled by default (as it is now) and
> the skin would disable the effect slots at start. In this case, the
> controller wouldn't have to touch the unit's enabled control.
>
> However...
>
> AFAIK no controllers are designed for chain enable switches separate from
> deck assignment and effect enable switches.
>
> Hm good point -- nevermind then. The inconsistency of enabling the effect
> chain slot by default and disabling the effect slot by default is fine w/
> me.
>
> How about syncing the state of the chain with the individual units?
> Switching the unit off switches all fx off. And whenever one or more
> effects are on, the chain is on
>
> This would be nice but it's surprisingly complicated to keep multiple
> controls in sync code-wise and it's easy to create infinite loops. Since
> the control system isn't transactional we can't update multiple controls
> atomically, and when you change one, it can run change handlers that go and
> change the state of the world on you before you update the other one to be
> in sync.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1103#issuecomment-270501047>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4hGM0VO_wdWim4SJp1Id4U-WizsHks5rPBg7gaJpZM4LY4lb>
> .
>

timrae commented:
> Hm good point -- nevermind then.

So the conclusion is that we should probably remove (in a separate PR) the chain enable switches since they don't offer any functionality that isn't provided by the other switches, right?

Be-ing commented:
> So the conclusion is that we should probably remove (in a separate PR) the chain enable switches since they don't offer any functionality that isn't provided by the other switches, right?

Yes, they should be removed from skins. Removing them from C++ might not be a good idea because people may have mapped controllers to them.

timrae commented:
It shouldn't be hard to confirm that using grep

On 5 Jan. 2017 07:34, "Be" <notifications@github.com> wrote:

> So the conclusion is that we should probably remove (in a separate PR) the
> chain enable switches since they don't offer any functionality that isn't
> provided by the other switches, right?
>
> Yes, they should be removed from skins. Removing them from C++ might not
> be a good idea because people may have mapped controllers to them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1103#issuecomment-270506821>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACsA4mw0X10Z6zRh_-1d3gfMxk1KDhoOks5rPB5XgaJpZM4LY4lb>
> .
>

Be-ing commented:
The trouble is accounting for people who have made their own mappings that they have not posted or contributed to Mixxx. It's pretty likely people have used the point and click wizard to map those buttons on screen. Removing a user-facing CO from C++ should only be done in rare circumstances when there is a really good reason to do so, for example removing the old flanger controls from 2.0 when the current effects system was implemented.

timrae commented:
:+1:

rryan commented:
> SGTM. Should that be in a separate PR or together with this?

In general I think it's best to keep PRs granular so let's do it in another one.

rryan commented on res/skins/LateNight/effect_unit.xml:
```diff
@@ -93,11 +93,12 @@
                 </Connection>
                 <Connection>
-                  <ConfigKey persist="true"><Variable name="group"/>,enabled</ConfigKey>
+                  <!-- FIXME: Quick hack until effect chains are implemented. -->
```
Could you add a connection block for Effect1 through Effect4?

Be-ing replied:
I don't think that would be worth the trouble considering the whole file will have to be replaced imminently. This is just a quick hack to make it keep working like it has been.

Be-ing replied:
I'll work on LateNight if no one else steps up to do it for 2.1.

rryan replied:
Thanks, even so I prefer every individual commit or PR to master to be complete and not half-working pending future work or design discussion. It's a good policy, since it allows releasing 2.1 at any commit from master even if pending work doesn't get done.
 
Here I am considering the case of effect chains with multiple effects. If like me, you have a variety of effect chains in your personal branch defined with multiple effects, this prevents LateNight from working with them.

Be-ing replied:
My concern is what will happen if toggling the enabled switches for all effects in the chain when they are not all in sync with each other. I'll test it out.

Be-ing replied:
The behavior with all effects connected to the button is erratic. It only lights up when Effect3 is enabled, regardless of the state of Effect1 and Effect2. I haven't quite figured out the rules determining how it behaves when clicking the button, but it's confusing. IMO leaving it as just connected to Effect1 is better for now.

rryan replied:
Hm, ok -- thanks for trying.

Be-ing committed:
add support for initial values for persistent ControlObjects

This lets the constructor specify a fallback value if no valid value for
the CO is found in the config file.

Be-ing committed:
add support for initial values to persistent ControlPushButtons

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into initial_value_for_persistent_cos

Be-ing committed:
change ControlObject constructors to take a defaultValue rather than initialValue

Be-ing committed:
add test for persistent ControlObjects

Be-ing committed:
add another test condition to persistent CO test

Be-ing committed:
Merge branch 'initial_value_for_persistent_cos' into effect_enable_switches

daschuer commented on src/effects/effectslot.cpp:
```diff
@@ -161,11 +160,11 @@ void EffectSlot::loadEffect(EffectPointer pEffect) {
         }
 
-        foreach (EffectParameterSlotPointer pParameter, m_parameters) {
+        for (EffectParameterSlotPointer pParameter : m_parameters) {
```
Please use const auto&, to avoid a deep copy.

daschuer commented on src/effects/effectslot.cpp:
```diff
@@ -161,11 +160,11 @@ void EffectSlot::loadEffect(EffectPointer pEffect) {
         }
 
-        foreach (EffectButtonParameterSlotPointer pParameter, m_buttonParameters) {
+        for (EffectButtonParameterSlotPointer pParameter : m_buttonParameters) {
```
Please use ```const auto&```. to avoid a deep copy. 

As a rule of thump, you can always use use ```const auto&```, and if the compiler complains about ```const``` use the universal reference ```auto&&```. This allows the complier to use optimal access and avoids implicit conversion.

Exceptions are raw pointers and 32 bit values where the reference stuff is not required.   
If you need a non const loop for implicit shared QT containers, use the Qt foreach macro. 

Byt the way: 
const in this context means that the QSharedPointer is const. The pointer carried by a shared pointer is always non const.

Be-ing replied:
> const in this context means that the QSharedPointer is const. The pointer carried by a shared pointer is always non const.

Thanks for clarifying. I was a bit confused how those loops were still working with const.

Be-ing committed:
assign EffectUnits to corresponding decks by default

and make changes to default persistent

Be-ing committed:
fix effects being disabled by default

Be-ing commented:
Effects, EQs, and QuickEffects are working now, but one of the EffectSlot tests is failing. I'll fix that tomorrow.

Be-ing committed:
fix EQs and QuickEffects

Be-ing committed:
fix EffectSlotTest ControlsReflectSlotState

Be-ing committed:
small code cleanup

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -146,9 +145,7 @@ void EffectSlot::loadEffect(EffectPointer pEffect) {
-        // Enabled is a persistent property of the effect slot, not of the
-        // effect. Propagate the current setting to the effect.
-        pEffect->setEnabled(m_pControlEnabled->get() > 0.0);
+        pEffect->setEnabled(m_pControlEnabled->toBool());
```
Could you keep the comment?

Be-ing replied:
I don't really understand the comment as it was. The slot enabled status has never been persistent, unless I'm misunderstanding the context of the comment.

Be-ing replied:
Oh, I think I understand. Persistent across calls to loadEffect(), not persistent across restarts of Mixxx.

rryan replied:
Ah, yea I think this comment predates the concept of a persistent CO :). Feel free to reword.

Be-ing committed:
add clarifying comment

Be-ing committed:
remove empty <Connection>
